### PR TITLE
[Enable gcs actor scheduler 1/n] Gcs scheduler calls hybrid_policy and reports pending info to autoscaler

### PR DESCRIPTION
**Pull request by Chong-Li**
## Why are these changes needed?
I'm trying to enable the gcs-based scheduler's feature flag by default, and make sure it passes the entire CI pipeline.

This PR focuses on two things:
1) Make the gcs scheduler directly call the hybrid_policy, which makes the raylet & gcs-based schedulers behave in a similar way in most of the common cases.
2) Make the gcs scheduler work with autoscaler, which requires reporting the count pending actors in `gcs_actor_manager` to `gcs_resource_manager`'s `node_resource_usages_`.

Basically, those are the two major issues that introduce CI failures. There will be one or two following PRs to fix remaining issues, and a final PR to run the entire CI pipeline with gcs scheduler feature flag enabled.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
Hi @Chong-Li  nice to know we are trying to enable it by default. My bad that I'm not following the refactoring of GCS side scheduler for a while so I have some high-level questions here.

If we use cluster_scheduler in gcs, why do we still need gcs_actor_distribution? Should we refactor this part first?

**Comment by iycheng** [comment]
Another question is, can we get rid of `gcs_actor_distribution`? It seems the main function right now is protocol implementation.

Also any plan to unify raylet one and gcs based one?

**Comment by iycheng** [comment]
Also, I think we need benchmarks and testing first. I'll see it as two steps to go: 1) benchmark and make sure CI is passing; 2)  from our side, we'll bake the feature for a while and monitor it in the nightly test.

**Comment by Chong-Li** [comment]
>

**Comment by Chong-Li** [comment]
> Another question is, can we get rid of `gcs_actor_distribution`? It seems the main function right now is protocol implementation.
> 
> Also any plan to unify raylet one and gcs based one?

Hmm... we may need to make the `RayletBasedActorScheduler` and `GcsBasedActorScheduler` coexist, as @scv119 proposed before (there will be a load balancer to select which scheduler at runtime). But even so, we still need a way to differ these two (currently by inheritance from GcsActorScheduler), so I'm not sure what kind of unification is needed. @iycheng

**Comment by Chong-Li** [comment]
> Also, I think we need benchmarks and testing first. I'll see it as two steps to go: 1) benchmark and make sure CI is passing; 2) from our side, we'll bake the feature for a while and monitor it in the nightly test.

This PR has not resolved all CI failures with gcs scheduler enabled. It now only focuses on the two major issues. As I mentioned in the title, there will be following PRs and a final PR to run the entire CI pipeline with gcs scheduler enabled. Should the benchmarking be done after that?

**Comment by rkooo567** [comment]
I think doing benchmark first and desmonstrating the perf benefits could be helpful for us since it will give us clear benefit why this needs to be turned on!

**Comment by iycheng** [comment]
> Hmm... we may need to make the RayletBasedActorScheduler and GcsBasedActorScheduler coexist, as @scv119 proposed before (there will be a load balancer to select which scheduler at runtime).

I'm not sure about this one. I think we don't need two schedulers. The only difference should be: 1) raylet based one is decentralized one and all raylet schedule things by themselves. 2) gcs based one is a centralized one, and we make scheduling decision in the same place. 

So what we need is where to send the scheduling request to, do it within raylet or do it with gcs. Maybe I missed something here. Please let me know if it doesn't make sense. @scv119 @Chong-Li

**Comment by iycheng** [comment]
Also, do we plan to do the cleanup before this PR? I'm ok with either, but we need a clear plan about this otherwise we might just execute it in the middle and have a lot of dead code.

We need to make sure we don't have schedulers specific for GCS or Raylet. They should just be a schedule and the only difference is about where the scheduling happens.

**Comment by iycheng** [comment]
> This PR only focuses on the two major issues. I just enabled gcs scheduler offline and make sure the related a few tests (test_advanced.py, test_output.py, etc) passed. As I mentioned in the title, there will be following PRs and a final PR to run the entire CI pipeline online with gcs scheduler enabled. Should the benchmarking be done after 

Do you have a proposed plan for this one? What are the next several PRs?

**Comment by Chong-Li** [comment]
> Also, do we plan to do the cleanup before this PR? I'm ok with either, but we need a clear plan about this otherwise we might just execute it in the middle and have a lot of dead code.
> 
> We need to make sure we don't have schedulers specific for GCS or Raylet. They should just be a schedule and the only difference is about where the scheduling happens.

The desired scenario for gcs scheduler is that all actor scheduling are handled by gcs (which is the requirement for any centralized scheduler in general), so each of them sees the same and consistent resource view. One of the consequences (that I can think of immediately) is that gcs' cluster_resource_manager only cares about the normal task (out of gcs scheduler's control) resource consumption in the periodic resource reports from raylets. So it basically means we have to allow some differences between gcs and raylet schedulers. 

I agree with you that the gcs_actor_distribution can be potentially moved into the gcs_actor_scheduler. But I feel that the top priority is always making gcs scheduler passes all CI and nightly tests, and then we could clean up the code to make it look neat.

**Comment by Chong-Li** [comment]
> I think doing benchmark first and desmonstrating the perf benefits could be helpful for us since it will give us clear benefit why this needs to be turned on!

I think @jjyao and @iycheng have done the benchmarking before, and the scheduling latency of gcs scheduler is about half of the raylet one (if my memory is right...). If we're worried about any performance regression, I'm ok to see it either right now or after the refactoring of gcs scheduler. @rkooo567

**Comment by Chong-Li** [comment]
> > This PR only focuses on the two major issues. I just enabled gcs scheduler offline and make sure the related a few tests (test_advanced.py, test_output.py, etc) passed. As I mentioned in the title, there will be following PRs and a final PR to run the entire CI pipeline online with gcs scheduler enabled. Should the benchmarking be done after
> 
> Do you have a proposed plan for this one? What are the next several PRs? The title of this one says enable it by default which gives me the feeling that it's the big milestone one.

The title said it takes several steps (**[1/n]**) and this is the first one. The next PR would be fixing trivia like PG wildcard resources and the interaction with ray call (which only happens in some corner case tests). Mixing those things with this PR might make reviewers easily distracted, so I decided to split them. This following PR will not be huge.

**Comment by iycheng** [comment]
Hi, @Chong-Li  thanks for some clarification.
I think @rkooo567 is making a valid point here since what we've tested is a long time ago and things are changing. We probably want a better view of what's the status quo.

> The desired scenario for gcs scheduler is that all actor scheduling are handled by gcs (which is the requirement for any centralized scheduler in general), so each of them sees the same and consistent resource view. One of the consequences (that I can think of immediately) is that gcs' cluster_resource_manager only cares about the normal task (out of gcs scheduler's control) resource consumption in the periodic resource reports from raylets. So it basically means we have to allow some differences between gcs and raylet schedulers.

I don't think this is what I mean. Let me rephrase this (maybe it's not making sense and please let me know).

**The goal here is to unify the resource reporting and schedulers and we only need to maintain two shared components, schedulers and reporters.**

In terms of the scheduler, we don't have a GCS-based scheduler or raylet-based scheduler. We only have one scheduler used by GCS and raylet.  We need to kill one scheduler.

The only difference is about which node to schedule the task/actor. That said, everything goes into the cluster scheduler and the only difference is about which node will we send the scheduling request to. If you send all requests to one node, it's like a centralized scheduler where only one place is making decisions, which is good when the cluster is overloaded. If requests are sent to the local node, it's like a decentralized scheduler, which is good for throughput when the cluster is pretty free.

The thing I don't like about this PR is that it's adding more things (or fixing the things) we are trying to kill. I understand we might want to claim the gain first and do clean up later, and I'm ok with this. But please share more concrete ideas about how do you plan to unify the schedulers in the longer term and also about the timeline.

**Comment by iycheng** [comment]
> The title said it takes several steps ([1/n]) and this is the first one. The next PR would be fixing trivia like PG wildcard resources and the interaction with ray call (which only happens in some corner case tests). Mixing those things with this PR might make reviewers easily distracted, so I decided to split them. This following PR will not be huge.

I think there is a gap here. The title should match the content of this PR. Even if there are stacked PRs for a feature, we shouldn't just give the title name of the feature directly which is really confusing. For example, this title gives people the feeling it's turning the feature on.

A good practice is like what @scv119 did here (https://github.com/pulls?q=is%3Apr+author%3Ascv119+archived%3Afalse+%22%2Fn%22+)

**Comment by Chong-Li** [comment]
Considering code cleanup (making gcs and raylet schedulers sharing as many modules as possible), the long-term plan is:
- Remove `RayletBasedActorScheduler` and `GcsBasedActorScheduler`. Instead, handling different actor requests (being scheduled at raylet or gcs) in `GcsActorScheduler`. 
- `GcsActorScheduler` calls `ClusterTaskManager`, and then `ClusterTaskManager` calls `ClusterResourceScheduler` to select node. Besides, because `ClusterTaskManager` maintains the infeasible and ready queues, the pending queue in `GcsActorManager` should be removed. After this change, the `GcsActorScheduler` only manages the lifecycle of leasing (send/cancel lease and push creation task), while the queueing and scheduling (this PR is actually doing the same thing in a quicker way) are handled by `ClusterTaskManager`. If the leftover responsibility of `GcsActorScheduler` is simple enough, we may even remove it and reimplement the functions in `GcsActorManager`.
- At raylets, the lease requests (that have already been handled by gcs' `ClusterTaskManager`) can skip the `ClusterTaskManager` and directly go to the `LocalTaskManager`.
- Turn on the gcs scheduling feature and resolve any leftover failures.

Any suggestions? @iycheng @scv119 @rkooo567 @raulchen @wumuzi520
